kyo-data/shared/src/main/scala/kyo/SafeClassTag.scala

```diff
@@ -0,0 +1,113 @@
+package kyo
+
+import kyo.Maybe
+import kyo.internal.SafeClassTagMacro
+import scala.annotation.tailrec
+import scala.quoted.*
+
+opaque type SafeClassTag[A] >: SafeClassTag.Element = Class[?] | SafeClassTag.Element
+
+object SafeClassTag:
+    sealed trait Element
+    case class Union(elements: List[SafeClassTag[Any]])        extends Element
+    case class Intersection(elements: List[SafeClassTag[Any]]) extends Element
+
+    sealed trait Primitive extends Element
```

Comment:
I think this could also be encoded as a bitmask on a Long, which would be easier to compress (either for intersections or unions, but probably not both?). Not sure if it's worth it given the need for generic type support...

Reply:
I think in this case it wouldn't be worth it since comparing these static tags via `eq` or by checking the class pointer is cheap. If we used a `Long`, we'd need to worry about boxing as well.